### PR TITLE
AKCORE-129: Permit fetching when some records are AVAILABLE, even when maxInFlightMessage limit is exceeded

### PR DESCRIPTION
**Pull request by chirag-wadhwa5**
This PR contains a minor bug fix

Rference: [AKCORE-129](https://confluentinc.atlassian.net/jira/software/c/projects/AKCORE/boards/2050/backlog?customFilter=4089&selectedIssue=AKCORE-129)

[AKCORE-129]: https://confluentinc.atlassian.net/browse/AKCORE-129?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: core/src/main/java/kafka/server/SharePartitionManager.java [hunk 1]
```diff
@@ -157,13 +157,18 @@ private void maybeProcessFetchQueue() {
...
-                                0,
-                                partitionMaxBytes,
-                                Optional.empty()));
+                    // Fetching over a topic should be able to proceed if any one of the following 2 conditions are met:
+                    // 1. The fetch is to happen somewhere in between the record states cached in the share partition.
+                    //    This is because in this case we don't need to check for the partition limit for in flight messages
+                    // 2. If condition 1 is not true, then that means we will be fetching new records which haven't been cached before.
+                    //    In this case it is necessary to check if the partition limit for in flight messages has been reached.
+                    if (sharePartition.nextFetchOffset() != sharePartition.endOffset() || sharePartition.canAcquireMore()) {
+                    topicPartitionData.put(topicIdPartition, new FetchRequest.PartitionData(
```
Comment (AndrewJSchofield) [inline core/src/main/java/kafka/server/SharePartitionManager.java:166]: Tiny nit. This line needs indenting.

File: core/src/main/java/kafka/server/SharePartitionManager.java [hunk 2]
```diff
@@ -157,13 +157,18 @@ private void maybeProcessFetchQueue() {
...
-                                sharePartition.nextFetchOffset(),
-                                0,
-                                partitionMaxBytes,
-                                Optional.empty()));
+                    // Fetching over a topic should be able to proceed if any one of the following 2 conditions are met:
+                    // 1. The fetch is to happen somewhere in between the record states cached in the share partition.
+                    //    This is because in this case we don't need to check for the partition limit for in flight messages
+                    // 2. If condition 1 is not true, then that means we will be fetching new records which haven't been cached before.
+                    //    In this case it is necessary to check if the partition limit for in flight messages has been reached.
+                    if (sharePartition.nextFetchOffset() != sharePartition.endOffset() || sharePartition.canAcquireMore()) {
```
Comment (AndrewJSchofield) [inline core/src/main/java/kafka/server/SharePartitionManager.java:165]: I think this is along the right lines, but I'm not sure it's quite right.

If I understand correctly, `endOffset` is the SPEO and it points to the final record which has already been fetched, while `nextFetchOffset` would point to the next offset to fetch. So, I wonder whether this is essentially out by one. I might be confused, but I don't think this is quite achieving exactly what condition (1) requires.

**Review by AndrewJSchofield — CHANGES_REQUESTED** [review]
Thanks for the PR. You are totally along the right lines with this fix.

Please can we have at least one unit test that exercises this situation so we can be sure we don't break it with future changes.

File: core/src/main/java/kafka/server/SharePartitionManager.java (hunk 2 above)
Comment (chirag-wadhwa5) [reply core/src/main/java/kafka/server/SharePartitionManager.java:165]: Thanks for reviewing the PR. You are absolutely right, nextFetchOffset should point to endOffset + 1. I wonder why the scenario did not produce an error locally in this case. Anyways, I will push the changes in the next commit.

File: core/src/main/java/kafka/server/SharePartitionManager.java (hunk 1 above)
Comment (chirag-wadhwa5) [reply core/src/main/java/kafka/server/SharePartitionManager.java:166]: Thanks, will make the changes in the next commit

**Review by AndrewJSchofield — APPROVED** [review]

**Pull request merged**